# Answering Business Questions using SQL

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

## Overview of the Data

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ('table','view');

 * sqlite:///chinook.db
Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table'),
 ('ifs', 'view')]

In [3]:
%%sql
SELECT *
FROM customer
LIMIT 5;

 * sqlite:///chinook.db
Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4)]

In [4]:
%%sql
SELECT *
FROM employee
LIMIT 5;

 * sqlite:///chinook.db
Done.


[(1, 'Adams', 'Andrew', 'General Manager', None, '1962-02-18 00:00:00', '2016-08-14 00:00:00', '11120 Jasper Ave NW', 'Edmonton', 'AB', 'Canada', 'T5K 2N1', '+1 (780) 428-9482', '+1 (780) 428-3457', 'andrew@chinookcorp.com'),
 (2, 'Edwards', 'Nancy', 'Sales Manager', 1, '1958-12-08 00:00:00', '2016-05-01 00:00:00', '825 8 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 2T3', '+1 (403) 262-3443', '+1 (403) 262-3322', 'nancy@chinookcorp.com'),
 (3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com'),
 (4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com')]

## Selecting New Albums to Purchase

In [5]:
%%sql
WITH usa AS
        (
        SELECT *
        FROM customer
        WHERE country = 'USA'
        ),
     usa_track AS
        (
        SELECT t.track_id,
               g.name
        FROM usa
        INNER JOIN invoice i ON usa.customer_id = i.customer_id
        INNER JOIN invoice_line il ON i.invoice_id = il.invoice_id
        INNER JOIN track t ON il.track_id = t.track_id
        INNER JOIN genre g ON t.genre_id = g.genre_id
        ),
     track_sold AS
        (
        SELECT name AS genre_name, 
               COUNT(*) AS number_of_tracks
        FROM usa_track
        GROUP BY 1
        ORDER BY 1)
        
SELECT genre_name,
       number_of_tracks,
       ROUND((CAST(number_of_tracks AS FLOAT) / 1051) * 100, 2) AS percentage_of_tracks
FROM track_sold
ORDER BY 3 DESC;

 * sqlite:///chinook.db
Done.


[('Rock', 561, 53.38),
 ('Alternative & Punk', 130, 12.37),
 ('Metal', 124, 11.8),
 ('R&B/Soul', 53, 5.04),
 ('Blues', 36, 3.43),
 ('Alternative', 35, 3.33),
 ('Latin', 22, 2.09),
 ('Pop', 22, 2.09),
 ('Hip Hop/Rap', 20, 1.9),
 ('Jazz', 14, 1.33),
 ('Easy Listening', 13, 1.24),
 ('Reggae', 6, 0.57),
 ('Electronica/Dance', 5, 0.48),
 ('Classical', 4, 0.38),
 ('Heavy Metal', 3, 0.29),
 ('Soundtrack', 2, 0.19),
 ('TV Shows', 1, 0.1)]

From the table above, we can see the top genres in USA. From what artists we have available to add to our stores, we have to add three.

We should add Artists:
1. *Red Tone* with Genre *Punk* which is having 12.37% share,
2. *Slim Jim Bites* with Genre *Blues* which is having 3.43% share,
3. *Meteor and the Girls* with Genre *Pop* which is having 2.09% share.

## Analyzing Employee Sales Performance

In [6]:
%%sql
WITH agent AS
        (
        SELECT *
        FROM employee
        WHERE title = 'Sales Support Agent'
        ),
     sales AS
        (
        SELECT *
        FROM agent
        INNER JOIN customer c ON agent.employee_id = c.support_rep_id
        ),
     performance AS
        (
        SELECT *
        FROM sales
        INNER JOIN invoice i ON sales.customer_id = i.customer_id
        )
SELECT employee_id AS agent_id,
       first_name || ' ' || last_name AS agent_name,
       SUM(total) AS total_sales,
       hire_date
FROM performance
GROUP BY 1
ORDER BY 1;

 * sqlite:///chinook.db
Done.


[(3, 'Jane Peacock', 1731.5100000000039, '2017-04-01 00:00:00'),
 (4, 'Margaret Park', 1584.0000000000034, '2017-05-03 00:00:00'),
 (5, 'Steve Johnson', 1393.920000000002, '2017-10-17 00:00:00')]

Above table tells us that *Jane Peacock* has the *most* sales and *Steve Johnson* has the *least* but important factor in this is related to their *hiring dates*.

## Analyzing Sales by Country

In [7]:
%%sql
WITH country_group AS
                (
                SELECT
                       country, 
                       COUNT(*) AS num_of_customers
                FROM customer
                GROUP BY 1
                ),
     invoice_group AS
                (
                SELECT 
                       billing_country,
                       COUNT(*) orders,
                       SUM(total) AS total
                FROM invoice
                GROUP BY 1
                ),
     country_invoice_group AS
                (
                SELECT 
                       CASE
                            WHEN num_of_customers = 1 THEN 'Other'
                            ELSE country
                       END AS country,
                       num_of_customers,
                       orders, 
                       total
                FROM country_group AS c
                INNER JOIN invoice_group AS i ON c.country = i.billing_country
                ),
     final_table AS
                (
                SELECT 
                       country,
                       SUM(num_of_customers) AS num_of_customers,
                       SUM(orders) AS orders,
                       SUM(total) AS total,
                       CASE
                           WHEN country = 'Other' THEN 1
                           ELSE 0
                       END AS sort
                FROM country_invoice_group
                GROUP BY 1
                ORDER BY sort ASC
                )
SELECT country,
       num_of_customers AS customers,
       total AS total_sales,
       total / num_of_customers AS average_sales,
       total / orders AS average_order_value
FROM final_table
ORDER BY sort ASC, total_sales DESC;

 * sqlite:///chinook.db
Done.


[('USA', 13, 1040.4899999999998, 80.0376923076923, 7.942671755725189),
 ('Canada', 8, 535.5900000000001, 66.94875000000002, 7.047236842105265),
 ('Brazil', 5, 427.68000000000006, 85.53600000000002, 7.0111475409836075),
 ('France', 5, 389.0699999999999, 77.81399999999998, 7.781399999999998),
 ('Germany', 4, 334.62, 83.655, 8.161463414634147),
 ('Czech Republic', 2, 273.24000000000007, 136.62000000000003, 9.108000000000002),
 ('United Kingdom', 3, 245.52, 81.84, 8.768571428571429),
 ('Portugal', 2, 185.13000000000002, 92.56500000000001, 6.383793103448276),
 ('India', 2, 183.14999999999998, 91.57499999999999, 8.72142857142857),
 ('Other', 15, 1094.9399999999998, 72.996, 7.4485714285714275)]

Based on the data, there may be opportunity in the following countries:

- Czech Republic
- United Kingdom
- India

It's worth keeping in mind that because the amount of data from each of these countries is relatively low.  Because of this, we should be cautious spending too much money on new marketing campaigns, as the sample size is not large enough to give us high confidence.  A better approach would be to run small campaigns in these countries, collecting and analyzing the new customers to make sure that these trends hold with new customers.

## Albums vs Individual Tracks

In [8]:
%%sql
SELECT il.invoice_id, t.track_id, a.album_id
FROM invoice_line il
INNER JOIN track t ON il.track_id = t.track_id
INNER JOIN album a ON t.album_id = a.album_id;

 * sqlite:///chinook.db
Done.


[(1, 1158, 91),
 (1, 1159, 91),
 (1, 1160, 91),
 (1, 1161, 91),
 (1, 1162, 91),
 (1, 1163, 91),
 (1, 1164, 91),
 (1, 1165, 91),
 (1, 1166, 91),
 (1, 1167, 91),
 (1, 1168, 91),
 (1, 1169, 91),
 (1, 1170, 91),
 (1, 1171, 91),
 (1, 1172, 91),
 (1, 1173, 91),
 (2, 3476, 322),
 (2, 482, 39),
 (2, 2701, 218),
 (2, 1641, 134),
 (2, 1119, 73),
 (2, 2324, 190),
 (2, 201, 20),
 (2, 819, 66),
 (2, 392, 34),
 (2, 2651, 215),
 (3, 2516, 203),
 (3, 2646, 214),
 (4, 3448, 314),
 (4, 2560, 207),
 (4, 3336, 260),
 (4, 829, 67),
 (4, 1872, 153),
 (4, 748, 58),
 (4, 1778, 146),
 (4, 2514, 203),
 (5, 1986, 163),
 (5, 1987, 163),
 (5, 1988, 163),
 (5, 1989, 163),
 (5, 1990, 163),
 (5, 1991, 163),
 (5, 1992, 163),
 (5, 1993, 163),
 (5, 1994, 163),
 (5, 1995, 163),
 (5, 1996, 163),
 (5, 1997, 163),
 (5, 1998, 163),
 (5, 1999, 163),
 (5, 2000, 163),
 (5, 2001, 163),
 (5, 2002, 163),
 (6, 1045, 83),
 (6, 30, 5),
 (7, 3477, 322),
 (7, 3439, 305),
 (7, 1102, 87),
 (7, 711, 55),
 (7, 1988, 163),
 (7, 3024, 239),
 (7, 834, 67),
 (7, 2596, 211),
 (7, 1343, 106),
 (7, 2644, 214),
 (7, 42, 6),
 (8, 2231, 183),
 (8, 2509, 203),
 (8, 81, 9),
 (8, 1758, 145),
 (8, 1455, 118),
 (8, 2946, 233),
 (8, 574, 46),
 (8, 1752, 144),
 (8, 2263, 185),
 (8, 1560, 125),
 (9, 206, 21),
 (9, 3456, 321),
 (9, 1041, 83),
 (9, 1279, 101),
 (9, 2098, 173),
 (9, 3399, 271),
 (9, 1983, 162),
 (9, 993, 79),
 (9, 196, 20),
 (10, 2663, 215),
 (10, 3290, 257),
 (11, 2597, 211),
 (11, 2203, 181),
 (11, 2420, 197),
 (11, 2643, 214),
 (11, 3072, 243),
 (11, 1138, 89),
 (11, 2391, 195),
 (11, 2257, 185),
 (11, 610, 49),
 (11, 1419, 115),
 (11, 3452, 318),
 (12, 1463, 118),
 (12, 690, 54),
 (12, 2663, 215),
 (12, 92, 10),
 (13, 2553, 206),
 (14, 541, 42),
 (15, 3050, 241),
 (15, 1489, 120),
 (15, 807, 65),
 (15, 3336, 260),
 (16, 2160, 178),
 (16, 875, 70),
 (16, 83, 9),
 (16, 16, 4),
 (16, 1756, 145),
 (16, 3460, 321),
 (16, 93, 10),
 (16, 24, 5),
 (16, 21, 4),
 (17, 2400, 195),
 (17, 1610, 131),
 (17, 1973, 162),
 (17, 2553, 206),
 (17, 59, 7),
 (17, 2270, 185),
 (17, 55, 7),
 (17, 2564, 207),
 (17, 2171, 179),
 (17, 2522, 203),
 (17, 1419, 115),
 (18, 1027, 82),
 (18, 3142, 141),
 (18, 1583, 127),
 (18, 1752, 144),
 (18, 2736, 221),
 (19, 105, 11),
 (19, 2669, 216),
 (19, 1784, 146),
 (19, 3460, 321),
 (19, 2640, 214),
 (19, 1310, 103),
 (19, 2510, 203),
 (20, 905, 72),
 (20, 102, 11),
 (20, 3336, 260),
 (20, 1132, 88),
 (20, 337, 30),
 (20, 60, 7),
 (20, 3357, 267),
 (20, 3225, 252),
 (20, 3297, 257),
 (20, 3297, 257),
 (21, 1461, 118),
 (21, 2556, 207),
 (21, 1434, 116),
 (21, 3135, 141),
 (21, 1473, 119),
 (21, 1502, 121),
 (21, 13, 1),
 (21, 2971, 235),
 (21, 2561, 207),
 (22, 1000, 80),
 (22, 383, 33),
 (23, 1, 1),
 (23, 6, 1),
 (23, 7, 1),
 (23, 8, 1),
 (23, 9, 1),
 (23, 10, 1),
 (23, 11, 1),
 (23, 12, 1),
 (23, 13, 1),
 (23, 14, 1),
 (24, 1146, 90),
 (24, 1147, 90),
 (24, 1148, 90),
 (24, 1149, 90),
 (24, 1150, 90),
 (24, 1151, 90),
 (24, 1152, 90),
 (24, 1153, 90),
 (24, 1154, 90),
 (24, 1155, 90),
 (24, 1156, 90),
 (24, 1157, 90),
 (25, 46, 6),
 (25, 948, 75),
 (25, 2185, 180),
 (25, 3049, 241),
 (25, 1167, 91),
 (25, 32, 5),
 (25, 1800, 147),
 (25, 1477, 119),
 (26, 2670, 216),
 (26, 85, 10),
 (27, 3439, 305),
 (27, 3014, 238),
 (27, 1280, 101),
 (27, 1496, 121),
 (27, 2229, 183),
 (27, 1445, 117),
 (27, 2230, 183),
 (27, 3105, 246),
 (28, 36, 5),
 (28, 3456, 321),
 (28, 932, 74),
 (29, 1025, 82),
 (29, 2462, 200),
 (29, 3341, 261),
 (29, 1626, 132),
 (29, 2395, 195),
 (29, 479, 39),
 (30, 919, 73),
 (30, 3461, 321),
 (30, 33, 5),
 (30, 2341, 191),
 (30, 1139, 89),
 (30, 3109, 246),
 (30, 2013, 164),
 (30, 990, 79),
 (30, 2413, 196),
 (30, 2396, 195),
 (30, 3464, 321),
 (31, 2731, 221),
 (31, 2732, 221),
 (31, 2733, 221),
 (31, 2734, 221),
 (31, 2735, 221),
 (31, 2736, 221),
 (31, 2737, 221),
 (31, 2738, 221),
 (31, 2739, 221),
 (31, 2740, 221),
 (31, 2741, 221),
 (31, 2742, 221),
 (31, 2743, 221),
 (31, 2744, 221),
 (31, 2745, 221),
 (31, 2746, 221),
 (31, 2747, 221),
 (31, 27

In [9]:
%%sql
WITH invoice_first_track AS
    (
     SELECT
         il.invoice_id invoice_id,
         MIN(il.track_id) first_track_id
     FROM invoice_line il
     GROUP BY 1
    )

SELECT
    album_purchase,
    COUNT(invoice_id) number_of_invoices,
    CAST(count(invoice_id) AS FLOAT) / (
                                         SELECT COUNT(*) 
                                                FROM invoice
                                      ) AS percent
FROM
    (
    SELECT
        ifs.*,
        CASE
            WHEN
                 (
                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 

                  EXCEPT 

                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id
                 ) IS NULL
             AND
                 (
                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id

                  EXCEPT 

                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 
                 ) IS NULL
             THEN "yes"
             ELSE "no"
         END AS "album_purchase"
     FROM invoice_first_track ifs
    )
GROUP BY album_purchase;

 * sqlite:///chinook.db
Done.


[('no', 500, 0.8143322475570033), ('yes', 114, 0.18566775244299674)]

Album purchases account for 18.6% of purchases.  Based on this data, I would recommend against purchasing only select tracks from albums from record companies, since there is potential to lose one fifth of revenue.

## Artist in Playlist

In [10]:
%%sql
SELECT 
        ar.name AS 'Artist Name', 
        COUNT(pt.playlist_id) AS 'Number of Playlists'
FROM playlist_track pt
INNER JOIN track t ON pt.track_id = t.track_id
INNER JOIN album a ON t.album_id = a.album_id
INNER JOIN artist ar ON a.artist_id = ar.artist_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10; 

 * sqlite:///chinook.db
Done.


[('Iron Maiden', 516),
 ('U2', 333),
 ('Metallica', 296),
 ('Led Zeppelin', 252),
 ('Deep Purple', 226),
 ('Lost', 184),
 ('Pearl Jam', 177),
 ('Eric Clapton', 145),
 ('Faith No More', 145),
 ('Lenny Kravitz', 143)]

We can see that `Iron Maiden` is the artist that has been used in most of the playlists.

## Tracks Purchased vs Tracks Not Purchased

In [29]:
%%sql
SELECT 
        COUNT(DISTINCT(il.track_id)) AS 'Tracks Purchased', 
        COUNT(DISTINCT(t.track_id)) - COUNT(DISTINCT(il.track_id)) AS 'Tracks Not Puchased'
FROM track t
LEFT JOIN invoice_line il ON t.track_id = il.track_id;

 * sqlite:///chinook.db
Done.


[(1806, 1697)]

We can see that out of `3503` tracks `1806` which is approx. `51.5%` have been purchased.